### PR TITLE
ci(auth): refactor to use Firebase Auth emulator

### PR DESCRIPTION
**Pull request by Salakar**
## Description

Depends on #5940. This PR adds the various Firebase Auth emulator http apis as testing utilities and adjusts the tests to now use the Firebase Auth emulator.

A follow up PR will look at improving the current tests, fixing or replacing tests that no longer work on the emulator and adding new tests that are now possible as a result of now using the emulator.

## Related Issues

*Replace this paragraph with a list of issues related to this PR from the [issue database](https://github.com/flutter/flutter/issues). Indicate, which of these issues are resolved or fixed by this PR. Note that you'll have to prefix the issue numbers with flutter/flutter#.*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/firebase_auth/firebase_auth/example/test_driver/test_utils.dart [hunk 1]
```diff
@@ -1,70 +1,215 @@
 // @dart = 2.9
```
Comment (mikehardy) [inline packages/firebase_auth/firebase_auth/example/test_driver/test_utils.dart:1]: just noting that you purged that `// @dart = 2.9` from other files but not from here, stands out

Comment (Salakar) [reply packages/firebase_auth/firebase_auth/example/test_driver/test_utils.dart:1]: I purged duplicates only, was 2 of them in the other file

Comment (Salakar) [reply packages/firebase_auth/firebase_auth/example/test_driver/test_utils.dart:1]: Note to self though; migrate tests to null safety (in follow up PR)

File: packages/firebase_auth/firebase_auth/example/test_driver/test_utils.dart [hunk 2]
```diff
@@ -1,70 +1,215 @@
...
+      'Authorization': 'Bearer owner',
+    },
+  );
+  final responseBody = Map<String, dynamic>.from(jsonDecode(response.body));
+  final verificationCodes =
+      List<Map<String, dynamic>>.from(responseBody['verificationCodes']);
+  return verificationCodes.reversed.firstWhere(
+    (verificationCode) => verificationCode['phoneNumber'] == phoneNumber,
+    orElse: () => {'code': 'NOT_FOUND'},
+  )['code'];
```
Comment (mikehardy) [inline packages/firebase_auth/firebase_auth/example/test_driver/test_utils.dart:102]: well isn't that all concise, makes me want to go and scrub the RNFB javascript version of the dinosaur coding style 🙈

Comment (Salakar) [reply packages/firebase_auth/firebase_auth/example/test_driver/test_utils.dart:102]: +1 for Dart 😆

File: packages/firebase_auth/firebase_auth/example/test_driver/firebase_auth_e2e.dart [hunk 3]
```diff
@@ -4,29 +4,47 @@
...
+    }
+  });
 
-    // Configure the Auth test suite to use the Auth Emulator
-    // This may not be enabled until the test suite is ported to:
-    //  - have ability to create disabled users
-    //  - have ability to fetch OOB and SMS verification codes
-    // JS implementation to port to dart here: https://github.com/invertase/react-native-firebase/pull/4552/commits/4c688413cb6516ecfdbd4ea325103d0d8d8d84a8#diff-44ccd5fb03b0d9e447820032866f2494c5a400a52873f0f65518d06aedafe302
-    // await FirebaseAuth.instance.useEmulator('http://localhost:9099');
+  setUp(() async {
```
Comment (mikehardy) [inline packages/firebase_auth/firebase_auth/example/test_driver/firebase_auth_e2e.dart:29]: Just verifying this is run before each individual test? I found a few cross-test dependencies when I moved to per-test-reset-recreate so it was worthwhile I think. I believe that's what is happening here and if so +1

**Review by mikehardy — APPROVED** [review]
A couple of tiny comments but no nits even. Odd a test got flakier on the emulator and the codes aren't the same but without testing it locally myself I'll obviously go with your experience. As mentioned now I want to go scrub the RNFB version, and also add the custom token test back in with your JWT generation method. Nice!

File: packages/firebase_auth/firebase_auth/example/test_driver/firebase_auth_e2e.dart (hunk 3 above)
Comment (Salakar) [reply packages/firebase_auth/firebase_auth/example/test_driver/firebase_auth_e2e.dart:29]: Aye this is before each test, [`setUp`](https://api.flutter.dev/flutter/flutter_test/setUp.html) vs [`setUpAll`](https://api.flutter.dev/flutter/flutter_test/setUpAll.html)

**Comment by Salakar** [comment]
> Odd a test got flakier on the emulator and the codes aren't the same

@mikehardy it was odd but I did repro with and without emulator to make sure, codes come through as `internal-error` on the emulator on those tests vs the ones we expected 😞 going to take a deep dive in the next PR, maybe it's some kind of nested exception when on the emulator and need to get the code elsewhere

**Pull request merged**